src/System.Text.Utf8/src/System/Text/Utf16/Utf16LittleEndianEncoder.cs

```diff
@@ -101,5 +101,28 @@ public static bool TryEncodeCodePoint(UnicodeCodePoint codePoint, ByteSpan buffe
             }
             return true;
         }
+
+        // TODO: Should we rewrite this to not use char.ConvertToUtf32 or is it fast enough?
+        public static bool TryDecodeCodePointFromString(string s, int index, out UnicodeCodePoint codePoint, out int encodedChars)
+        {
+            if (index < 0 || index >= s.Length)
+            {
+                codePoint = default(UnicodeCodePoint);
+                encodedChars = 0;
+                return false;
+            }
+
+            if (index == s.Length - 1 && char.IsSurrogate(s[index]))
+            {
+                codePoint = default(UnicodeCodePoint);
+                encodedChars = 0;
+                return false;
+            }
+
+            encodedChars = char.IsHighSurrogate(s[index]) ? 2 : 1;
+            codePoint = (UnicodeCodePoint)(unchecked((uint)char.ConvertToUtf32(s, index)));
```

Comment:
This call can throw in the case where you have a high surrogate followed by another high surrogate, I think.